File: examples/latest/navigate/ios/TruckGuidance/TruckGuidance/TruckRestrictionView.swift
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import SwiftUI

// A model class to be used for data binding. The example class will use this model to update
// data while traversing along a route. The data will be shown by the TruckRestrictionView panel.
class TruckRestrictionModel: ObservableObject {
    @Published var restrictionDescription: String?
    @Published var isViewVisible: Bool
    
    init(restrictionDescription: String = "n/a",
         isViewVisible: Bool = false) {
        self.restrictionDescription = restrictionDescription
        self.isViewVisible = isViewVisible
    }
}

// A simple view to show the next TruckRestrictionWarning event.
struct TruckRestrictionView: View {
    
    // The model which is updated by the example class when new data is provided by the VisualNavigator.
    @ObservedObject var model: TruckRestrictionModel
    
    // The width and height of the container view.
    let w: CGFloat = 125
    let h: CGFloat = 60
    
    // Margin for text positioning.
    private let margin: CGFloat = 8
    
    // Corner radius for rounded background.
    private let cornerRadius: CGFloat = 8.0
    
    // Custom background color.
    private let customBackgroundColor = Color(red: 18/255, green: 109/255, blue: 249/255)
    
    var body: some View {
        if model.isViewVisible {
            ZStack(alignment: .leading) {
                // Background rectangle with rounded corners.
                RoundedRectangle(cornerRadius: cornerRadius)
                    .fill(customBackgroundColor)
                    .frame(width: w, height: h)
                
                // Restriction description text, left-aligned and truncated if too long.
                if let restrictionDescription = model.restrictionDescription, !restrictionDescription.isEmpty {
                    Text(restrictionDescription)
                        .font(.system(size: 14, weight: .bold))
                        .foregroundColor(.white)
                        .lineLimit(1)
                        .truncationMode(.tail)
                        .padding(.leading, margin)
                        .frame(width: w - 2 * margin, height: h, alignment: .center)
                }
            }
            .frame(width: w, height: h)
            .background(Color.clear)
        }
    }
}

struct TruckRestrictionView_Previews: PreviewProvider {
    static var previews: some View {
        // Create a sample model for the preview.
        let sampleModel = TruckRestrictionModel(
            restrictionDescription: "n/a"
        )
        
        TruckRestrictionView(model: sampleModel)
            .previewLayout(.sizeThatFits)
    }
}

File: examples/latest/navigate/flutter/truck_guidance_app/lib/truck_restriction_view.dart
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import 'package:flutter/material.dart';

/// A custom widget that mimics TruckRestrictionView from Java.
/// It draws a rounded rectangle with a red stroke and centers a description inside.
class TruckRestrictionView extends StatefulWidget {
  String description = "";
  TruckRestrictionView({Key? key, required this.description}) : super(key: key);

  @override
  State<TruckRestrictionView> createState() => _TruckRestrictionViewState();
}

class _TruckRestrictionViewState extends State<TruckRestrictionView> {
  @override
  Widget build(BuildContext context) {
    // The Java code uses wInDP=160 and hInDP=40.
    return CustomPaint(
      size: const Size(160, 40),
      painter: TruckRestrictionViewPainter(description: widget.description),
    );
  }
}

class TruckRestrictionViewPainter extends CustomPainter {
  String description = "";
  TruckRestrictionViewPainter({required this.description});

  @override
  void paint(Canvas canvas, Size size) {
    if (description.isEmpty) return;

    double textSizeInDP = 15;
    Paint paint = Paint()..isAntiAlias = true;

    // Define the rectangle.
    double left = 0;
    double top = 0;
    double right = 160;
    double bottom = 40;
    double strokeWidth = 6;

    // Adjust rectangle for stroke.
    double innerLeft = left + strokeWidth / 2;
    double innerTop = top + strokeWidth / 2;
    double innerRight = right - strokeWidth / 2;
    double innerBottom = bottom - strokeWidth / 2;
    Rect innerRect = Rect.fromLTWH(innerLeft, innerTop, innerRight - innerLeft, innerBottom - innerTop);

    // Draw white background with rounded corners.
    paint.style = PaintingStyle.fill;
    paint.color = Colors.white;
    canvas.drawRRect(RRect.fromRectAndRadius(innerRect, const Radius.circular(10)), paint);

    // Draw red stroke.
    paint.style = PaintingStyle.stroke;
    paint.color = Colors.red;
    paint.strokeWidth = strokeWidth;
    canvas.drawRRect(RRect.fromRectAndRadius(innerRect, const Radius.circular(10)), paint);

    // Draw the description text centered.
    TextSpan span = TextSpan(
      style: TextStyle(color: Colors.black, fontSize: textSizeInDP, fontWeight: FontWeight.bold),
      text: description,
    );
    TextPainter tp = TextPainter(text: span, textAlign: TextAlign.center, textDirection: TextDirection.ltr);
    tp.layout();
    Offset textOffset = Offset(innerRect.center.dx - tp.width / 2, innerRect.center.dy - tp.height / 2);
    tp.paint(canvas, textOffset);
  }

  @override
  bool shouldRepaint(covariant TruckRestrictionViewPainter oldDelegate) {
    return oldDelegate.description != description;
  }
}
